### ISSUE TITLE
How to switch audio input devices?

### ISSUE THREAD

**Issue by spiffytech** [twilio/twilio-video.js#847]
I'm connecting to a video room and allowing twilio-video to select the audio input device. If the user is using a plug-in mic (e.g., a headset) and unplugs the mic, audio doesn't switchover to another input (e.g., their laptop built-in mic).

I can see how to use `getUserMedia` to list devices, but not how to connect the device to my local participant. What's the correct way to switch audio inputs on an active room?

**Comment by manjeshbhargav** [comment]
Hi @spiffytech ,

Thanks for writing in with your question. Please take a look at our [quickstart example for media devices](https://github.com/twilio/video-quickstart-js/blob/master/examples/mediadevices/src/helpers.js) and let me know if it helps. You can also build and run this example in your browser by following the instructions in the README at the root of this repository.

Please feel free to re-open this issue if you need any further clarifications.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by spiffytech** [comment]
Looking through that sample code, I built a handler for media device changes. Using some `console.log` debugging, when I unplug my headset I can see the handler select and apply my built-in mic, but I don't get any sound out the other end of the video connection.

```javascript
const mkDeviceChangeHandler = (room: Room) => async () => {
  const devices = await navigator.mediaDevices.enumerateDevices();
  const audioDevices = devices.filter(device => device.kind === 'audioinput');
  if (!audioDevices[0]) return;
  const track = await createLocalAudioTrack({deviceId: {exact: audioDevices[0].deviceId}});
  room.localParticipant.audioTracks.forEach(publication => {
    publication.track.stop();
    room.localParticipant.unpublishTrack(publication.track);
  });
  room.localParticipant.publishTrack(track);
};

navigator.mediaDevices.ondevicechange = mkDeviceChangeHandler(room);
```

**Comment by spiffytech** [comment]
I got it working - I think the above code is right, but the other end of the video connection wasn't correctly responding to audio track subscriptions/unsubscriptions.

**Comment by manjeshbhargav** [comment]
@spiffytech ,

It's great that you got the device change scenario to work! On the remote side, are you still having problems with subscribing to and hearing the new AudioTrack?

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by iawen28** [comment]
I had the same issue and found this thread very helpful. I didn't need to handle anything on the remote side, but got it working by appending and removing track.attach() and track.detach() to the local media html element (in addition to publishing and unpublishing tracks to the room).
